integration/client_test.go

```diff
@@ -0,0 +1,295 @@
+// +build integration
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	gcpcachev2 "github.com/envoyproxy/go-control-plane/pkg/cache/v2"
+	gcpserverv2 "github.com/envoyproxy/go-control-plane/pkg/server/v2"
+	gcpserverv3 "github.com/envoyproxy/go-control-plane/pkg/server/v3"
+	gcptest "github.com/envoyproxy/go-control-plane/pkg/test"
+	resourcev2 "github.com/envoyproxy/go-control-plane/pkg/test/resource/v2"
+	gcptestv2 "github.com/envoyproxy/go-control-plane/pkg/test/v2"
+	"github.com/stretchr/testify/assert"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	corev2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	"github.com/envoyproxy/xds-relay/internal/app/upstream"
+	"github.com/envoyproxy/xds-relay/internal/pkg/log"
+)
+
+const (
+	nodeID           = "node-id"
+	originServerPort = 18000
+	loglevel         = "fatal"
+)
+
+func TestXdsClientGetsIncrementalResponsesFromUpstreamServer(t *testing.T) {
+	updates := 2
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	testLogger := log.New(loglevel)
+
+	snapshotsv2, configv2 := createSnapshotCache(updates, testLogger)
+	cb := gcptestv2.Callbacks{Signal: make(chan struct{})}
+	respCh, _, err := setup(ctx, testLogger, snapshotsv2, configv2, &cb)
+	if err != nil {
+		assert.Fail(t, "Setup failed: %s", err.Error())
+		return
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(updates)
+	version := 0
+	go func() {
+		for {
+			select {
+			case r, more := <-respCh:
+				if !more {
+					return
+				}
+				assert.Equal(t, r.VersionInfo, "v"+strconv.Itoa(version))
+				version++
+				wg.Done()
+			}
+		}
+	}()
+
+	sendResponses(ctx, testLogger, updates, snapshotsv2, configv2)
+	wg.Wait()
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	select {
+	case <-timeoutCtx.Done():
+		assert.Fail(t, "request count did not match")
+	case <-time.After(1 * time.Second):
+		// There should be one extra waiting request for the watch
+		assert.Equal(t, updates+1, cb.Requests)
+		break
+	}
+}
+
+func TestXdsClientShutdownShouldCloseTheResponseChannel(t *testing.T) {
+	updates := 1
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	testLogger := log.New(loglevel)
+
+	snapshotsv2, configv2 := createSnapshotCache(updates, testLogger)
+	cb := gcptestv2.Callbacks{Signal: make(chan struct{})}
+	respCh, shutdown, err := setup(ctx, testLogger, snapshotsv2, configv2, &cb)
+	if err != nil {
+		assert.Fail(t, "Setup failed: %s", err.Error())
+		return
+	}
```

Comment:
Do you want to pull this out to a helper / setup function since it's used across a few of these tests?

Reply:
I think i have pulled the most of the common code into setup. I did not want to return 5 things out of setup and hence the repeated lines
```
        snapshotsv2, configv2 := createSnapshotCache(updates, testLogger)
	cb := gcptestv2.Callbacks{Signal: make(chan struct{})}
	respCh, shutdown, err := setup(ctx, testLogger, snapshotsv2, configv2, &cb)
```

Reply:
I have moved the logger out to a common place. I think the other common line is ctx creation, which is specific to each test.